### PR TITLE
Discount curve calibrator

### PR DESCRIPTION
**Pull request by yukiiwashita**
- Discount curve calibration to money markets and swaps, specific to ISDA standard model for CDS.

### CODE REVIEW

File: modules/market/src/main/java/com/opengamma/strata/market/curve/IsdaCreditCurveNode.java [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+   * Gets the date of this node within the curve.
+   * <p>
+   * Each curve node has an associated date which defines the x-value in the curve. 
+   * This is typically the adjusted end date of the instrument.
+   * 
+   * @param tradeDate  the trade date
+   * @param refData  the reference data
+   * @return the node date
+   */
+  public abstract LocalDate getNodeDate(LocalDate tradeDate, ReferenceData refData);
```
Comment (jodastephen) [inline modules/market/src/main/java/com/opengamma/strata/market/curve/IsdaCreditCurveNode.java]: should be `date()` to match `CurveNode`

File: modules/market/src/main/java/com/opengamma/strata/market/curve/DepositIsdaCreditCurveNode.java [hunk 2]
```diff
@@ -0,0 +1,688 @@
...
+   * This defines the term year fraction.
+   */
+  @PropertyDefinition(validate = "notNull")
+  private final DayCount dayCount;
+
+  //-------------------------------------------------------------------------
+  /**
+   * Returns a curve node for a term deposit.
+   * <p>
+   * {@code label} will be created using {@code tenor}.
```
Comment (jodastephen) [inline modules/market/src/main/java/com/opengamma/strata/market/curve/DepositIsdaCreditCurveNode.java]: "The label will be "...

File: modules/market/src/main/java/com/opengamma/strata/market/curve/IsdaCreditCurveNode.java (hunk 1 above)
Comment (jodastephen) [inline modules/market/src/main/java/com/opengamma/strata/market/curve/IsdaCreditCurveNode.java]: Javadoc would be "Calculates the date associated with the node."

File: modules/market/src/main/java/com/opengamma/strata/market/curve/SwapIsdaCreditCurveNode.java [hunk 3]
```diff
@@ -0,0 +1,771 @@
...
+   * The compounding is not allowed in this node. Thus the frequency is the same as the accrual periodic frequency.
+   */
+  @PropertyDefinition(validate = "notNull")
+  private final Frequency paymentFrequency;
+
+  //-------------------------------------------------------------------------
+  /**
+   * Returns a curve node for a standard fixed-Ibor swap.
+   * <p>
+   * {@code label} will be created from {@code tenor}.
```
Comment (jodastephen) [inline modules/market/src/main/java/com/opengamma/strata/market/curve/SwapIsdaCreditCurveNode.java]: The label

File: modules/pricer/src/main/java/com/opengamma/strata/pricer/credit/cds/IsdaCompliantDiscountCurveCalibrator.java [hunk 4]
```diff
@@ -0,0 +1,454 @@
...
+ * <p>
+ * Once calibrated, the curves are then available for use.
+ * Each node in the curve definition becomes a parameter in the matching output curve.
+ */
+public final class IsdaCompliantDiscountCurveCalibrator {
+
+  /**
+   * Default implementation.
+   */
+  public static final IsdaCompliantDiscountCurveCalibrator DEFAULT = new IsdaCompliantDiscountCurveCalibrator();
```
Comment (jodastephen) [inline modules/pricer/src/main/java/com/opengamma/strata/pricer/credit/cds/IsdaCompliantDiscountCurveCalibrator.java]: Should copy the design of `CurveCalibrator`. Have a `standard()` static method and a private constructor. Consider whether root finding parameters should be able to be passed in.

File: modules/pricer/src/main/java/com/opengamma/strata/pricer/credit/cds/IsdaCompliantDiscountCurveCalibrator.java [hunk 5]
```diff
@@ -0,0 +1,454 @@
...
+   * @param curveDcc  the curve day count
+   * @param name  the curve name
+   * @param currency  the currency
+   * @param marketData  the market data
+   * @param computeJacobian  the Jacobian matrices are computed if true
+   * @param refData  the reference data
+   * @return the ISDA compliant discount curve
+   */
+  public IsdaCompliantZeroRateDiscountFactors calibrate(
+      List<IsdaCreditCurveNode> curveNode,
```
Comment (jodastephen) [inline modules/pricer/src/main/java/com/opengamma/strata/pricer/credit/cds/IsdaCompliantDiscountCurveCalibrator.java]: curveNode -> curveNodes

File: modules/pricer/src/main/java/com/opengamma/strata/pricer/credit/cds/IsdaCompliantDiscountCurveCalibrator.java [hunk 6]
```diff
@@ -0,0 +1,454 @@
...
+      LocalDate curveValuationDate,
+      DayCount curveDcc,
+      CurveName name,
+      Currency currency,
+      MarketData marketData,
+      boolean computeJacobian,
+      ReferenceData refData) {
```
Comment (jodastephen) [inline modules/pricer/src/main/java/com/opengamma/strata/pricer/credit/cds/IsdaCompliantDiscountCurveCalibrator.java]: Add an `IsdaCreditCurveDefinition` class to capture these inputs, (market and reference data would still be passed in here)

File: modules/pricer/src/main/java/com/opengamma/strata/pricer/credit/cds/IsdaCompliantDiscountCurveCalibrator.java [hunk 7]
```diff
@@ -0,0 +1,454 @@
...
+        termDepositYearFraction[i] = termDeposit.getDayCount().relativeYearFraction(cvDateTmp, adjMatDate);
+        ArgChecker.isTrue(nTermDeposit == i, "TermDepositCurveNode should not be after FixedIborSwapCurveNode");
+        ++nTermDeposit;
+      } else if (node instanceof SwapIsdaCreditCurveNode) {
+        SwapIsdaCreditCurveNode swap = (SwapIsdaCreditCurveNode) node;
+        cvDateTmp = swap.getSpotDateOffset().adjust(curveSnapDate, refData);
+        curveNodeTime[i] = curveDcc.relativeYearFraction(cvDateTmp, adjMatDate);
+        BusinessDayAdjustment busAdj = swap.getBusinessDayAdjustment();
+        swapLeg[i] = new BasicFixedLeg(cvDateTmp, cvDateTmp.plus(swap.getTenor()), swap.getPaymentFrequency().getPeriod(),
+            swap.getDayCount(), curveDcc, busAdj, refData);
```
Comment (jodastephen) [inline modules/pricer/src/main/java/com/opengamma/strata/pricer/credit/cds/IsdaCompliantDiscountCurveCalibrator.java]: Format with each param on a new line

File: modules/pricer/src/main/java/com/opengamma/strata/pricer/credit/cds/IsdaCompliantDiscountCurveCalibrator.java [hunk 8]
```diff
@@ -0,0 +1,454 @@
...
+      if (i > 0) {
+        ArgChecker.isTrue(curveNodeTime[i] - curveNodeTime[i - 1] > 0, "curve nodes should be ascending in terms of tenor");
+        ArgChecker.isTrue(cvDateTmp.equals(curveSpotDate), "spot lag should be common for all of the curve nodes");
+      } else {
+        ArgChecker.isTrue(curveNodeTime[i] >= 0d, "the first node should be after curve spot date");
+        curveSpotDate = cvDateTmp;
+      }
+    }
+    ImmutableList<ParameterMetadata> parameterMetadata = paramMetadata.build();
+    CurveMetadata baseMetadata = DefaultCurveMetadata.builder()
```
Comment (jodastephen) [inline modules/pricer/src/main/java/com/opengamma/strata/pricer/credit/cds/IsdaCompliantDiscountCurveCalibrator.java]: @Use `Curves.zeroRates()` - can pass parameter metadata in too

File: modules/pricer/src/main/java/com/opengamma/strata/pricer/credit/cds/IsdaCompliantDiscountCurveCalibrator.java [hunk 9]
```diff
@@ -0,0 +1,454 @@
...
+      }
+      for (int i = nTermDeposit; i < nNode; ++i) {
+        sensiTotal[i] = sensiSwap.rowArray(i - nTermDeposit);
+      }
+      return DoubleMatrix.ofUnsafe(sensiTotal);
+    }
+    return DoubleMatrix.EMPTY;
+  }
+
+  private DoubleArray sensitivityDeposit(InterpolatedNodalCurve curve, double termDepositYearFraction, int index,
```
Comment (jodastephen) [inline modules/pricer/src/main/java/com/opengamma/strata/pricer/credit/cds/IsdaCompliantDiscountCurveCalibrator.java]: Params on new lines

File: modules/pricer/src/main/java/com/opengamma/strata/pricer/credit/cds/IsdaCompliantDiscountCurveCalibrator.java [hunk 10]
```diff
@@ -0,0 +1,454 @@
...
+  }
+
+  //-------------------------------------------------------------------------
+  /* crude swap fixed leg description */
+  private final class BasicFixedLeg {
+    private final int nPayment;
+    private final double[] swapPaymentTime;
+    private final double[] yearFraction;
+
+    public BasicFixedLeg(LocalDate curveSpotDate, LocalDate maturityDate, Period swapInterval, DayCount swapDCC,
```
Comment (jodastephen) [inline modules/pricer/src/main/java/com/opengamma/strata/pricer/credit/cds/IsdaCompliantDiscountCurveCalibrator.java]: Params on new lines

File: modules/pricer/src/main/java/com/opengamma/strata/pricer/credit/cds/IsdaCompliantDiscountCurveCalibrator.java (hunk 8 above)
Comment (yukiiwashita) [inline modules/pricer/src/main/java/com/opengamma/strata/pricer/credit/cds/IsdaCompliantDiscountCurveCalibrator.java]: Curves.zeroRates(CurveName, DayCount) is used. 
The parameter metadata list is kept separated for the latter purpose.

**Pull request merged**